executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -174,46 +175,111 @@ impl<'a, T: FieldElement> BlockMachineProcessor<'a, T> {
                 .format("\n  ");
             format!("Code generation failed: {shortened_error}\nRun with RUST_LOG=trace to see the code generated so far.")
         })?;
-        self.check_block_shape(&result.code)?;
+
+        // During block shape checking, we allow removing assignments
+        // that would otherwise destroy the block shape, as long as these assignments
+        // are not required to compute the requested variables.
+        let optional_vars = code_cleaner::optional_vars(&result.code, &requested_known);
+        let ItemsToRemove {
+            vars_to_remove,
+            machine_calls_to_remove,
+        } = self.check_block_shape(&result.code, &optional_vars)?;
+        result.code = code_cleaner::remove_variables(result.code, vars_to_remove);
+        result.code = code_cleaner::remove_machine_calls(result.code, &machine_calls_to_remove);
+
         Ok((result, prover_functions))
     }
 
     /// Verifies that each column and each bus send is stackable in the block.
     /// This means that if we have a cell write or a bus send in row `i`, we cannot
     /// have another one in row `i + block_size`.
-    fn check_block_shape(&self, code: &[Effect<T, Variable>]) -> Result<(), String> {
+    /// Returns a set of (unnecessary) variables and machine calls to remove to maintain
+    /// proper block shape.
+    fn check_block_shape(
```

Comment:
```suggestion
    fn enforce_valid_block_shape(
```

Maybe it could also consume the code and return a `Result<Vec<Effect<...>>, String`? Then the "enforce" term actually makes sense and the calling code wouldn't have to deal with the `ItemsToRemove`.